polkadot/node/network/statement-distribution/src/v2/mod.rs

```diff
@@ -2126,17 +2135,54 @@ async fn provide_candidate_to_grid<Context>(
 	}
 }
 
-fn group_for_para(
+// Utility function to populate per relay parent `ParaId` to `GroupIndex` mappings.
+fn determine_groups_per_para(
 	availability_cores: &[CoreState],
 	group_rotation_info: &GroupRotationInfo,
-	para_id: ParaId,
-) -> Option<GroupIndex> {
-	// Note: this won't work well for on-demand parachains as it assumes that core assignments are
-	// fixed across blocks.
-	let core_index = availability_cores.iter().position(|c| c.para_id() == Some(para_id));
+	maybe_claim_queue: &Option<ClaimQueueSnapshot>,
+	max_candidate_depth: usize,
+) -> HashMap<ParaId, Vec<GroupIndex>> {
+	// Determine the core indices occupied by each para at the current relay parent. To support
+	// on-demand parachains we also consider the core indices at next block if core has a candidate
+	// pending availability.
+	let para_core_indices = availability_cores.iter().enumerate().filter_map(|(index, core)| {
+		match core {
+			CoreState::Scheduled(scheduled_core) =>
+				Some((scheduled_core.para_id, CoreIndex(index as u32))),
+			CoreState::Occupied(occupied_core) => {
+				if max_candidate_depth >= 1 {
+					// Use claim queue if available, or fallback to `next_up_on_available`
+					let maybe_scheduled_core = match maybe_claim_queue {
+						Some(claim_queue) => {
+							// What's up next on this core ?
+							fetch_next_scheduled_on_core(claim_queue, CoreIndex(index as u32))
+						},
+						None => {
+							// Runtime doesn't support claim queue runtime api. Fallback to
+							// `next_up_on_available`
+							occupied_core.next_up_on_available.clone()
+						},
+					};
+
+					maybe_scheduled_core
+						.filter(|scheduled_core| scheduled_core.para_id == occupied_core.para_id())
```

Comment:
Feels like this doesn't work if we change cores for paras for example:
Block1:
Core0, Occupied(Para1)
Core1, Occupied(Para2)
Core2, Occupied(Para1)


ClaimQueue: Core0 -> Para2, Core1 -> Para1, Core2 -> Para1

Reply:
This filter does exactly that, it checks that the upcoming assignment is of the same para, otherwise the para / core tuple is filtered out.

Reply:
Do you refer strictly to Core1, where we should consider it as well ? If so I have to think about that a bit.

Reply:
Well, since we are interested in just the backing groups for this relay parent only, I don't think we should bother about the case I mentioned above.

Reply:
This is not perfect, since for example in your case, para1 could still have assigned some other cores like Core10, and the backing group for core0 would still be used to back it and it would still be backed at Core10, but that also assumes some candidate pending availability there is included, which complicates things quite a lot

Reply:
I removed the filter, based on your example we now get backing groups for Core1, Core2 backing groups for Para1.